### PR TITLE
feat(ottl): Add a new ottl trim function that trims leading and trailing whitespace from a string

### PR DESCRIPTION
**Pull request by rnishtala-sumo**
#### Description

A field test exists in an event, which contains the string " this is a test ".

I would like to be able to transform this using an ottl function trim, that I would define as trim(attribute["test"]), which would trim this down to "this is a test".

- Trim(" this is a test ") results in `this is a test`

Link to the issue - https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34100

### CODE REVIEW

**Review by TylerHelmuth — COMMENTED** [review]
Please add an entry to `ottlfuncs/README` for this new converter

**Review by dmitryax — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
@rnishtala-sumo can you add to the description the issue this PR closes

File: pkg/ottl/ottlfuncs/func_trim.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type TrimArguments[K any] struct {
+	Target ottl.StringGetter[K]
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_trim.go:15]: Lets add an optional parameter that allows specifying what to trim.  I believe it should be a string and the converter will trim that string from both sides of the target string.

File: .chloggen/ottl-trim-function.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: 'enhancement'
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: ottl
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add a new ottl trim function that trims leading and trailing whitespace from a string.
```
Comment (edmocosta) [inline .chloggen/ottl-trim-function.yaml:10]: I'd update this note given the function is not limited to whitespaces.

File: pkg/ottl/ottlfuncs/func_trim.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+func trim[K any](target ottl.StringGetter[K], replacement ottl.Optional[string]) ottl.ExprFunc[K] {
+	return func(ctx context.Context, tCtx K) (any, error) {
+		replacementString := replacement.Get()
+		val, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+		if replacementString == "" {
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_trim.go:40]: If the `replacement` argument is optional, we could use the `.IsEmpty()` function to check whether a value was provided or not. 

I'd suggest changing it a bit making it similar to the `strings.Trim` function, meaning that the default whitespace replacement would only be used when `replacement.IsEmpty()`, and `!replacement.IsEmpty() && replacement.Get() == ""` would return the original string. 

Any thoughts @TylerHelmuth?

File: pkg/ottl/ottlfuncs/README.md [hunk 4]
```diff
@@ -1780,9 +1780,21 @@ The `Split` Converter separates a string by the delimiter, and returns an array
 
 If the `target` is not a string or does not exist, the `Split` Converter will return an error.
 
+### Trim
+
+```Trim(target, replacement)```
+
+The `Trim` Converter removes the leading and trailing character (default: whitespace).
```
[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/README.md:1787]: Yes. Take a look in this README how other Optional parameters are indicated (can use flatten as an example).

File: pkg/ottl/ottlfuncs/func_trim.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+func trim[K any](target ottl.StringGetter[K], replacement ottl.Optional[string]) ottl.ExprFunc[K] {
+	return func(ctx context.Context, tCtx K) (any, error) {
+		replacementString := replacement.Get()
+		val, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+		if replacement.IsEmpty() || replacementString == "" {
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_trim.go:40]: Above line 34, lets to a check for if `replacement == ""` and error if it is.  Calling `Trim(name, "")` implies a nop, but we'd be trimming `" "`.

Comment (rnishtala-sumo) [reply pkg/ottl/ottlfuncs/func_trim.go:40]: @TylerHelmuth isn't an empty replacement string expected since the argument is optional?

Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_trim.go:40]: I'm saying we shouldn't allow a user to set the optional parameter to be "".

Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_trim.go:40]: IsEmpty returns false if the user sets the optional parameter to any value I thought, even default values like "".

Comment (rnishtala-sumo) [reply pkg/ottl/ottlfuncs/func_trim.go]: Yes, `replacement.isEmpty()` returns false with a default value of `""` for the `replacementString`.

However, the replacementString is set to `""` under two conditions

- When the replacement isn't provided (uses default `""`)
- When the `replacementString` is explicitly set to `""`

If we return an error when the `replacementString` is set to `""`, we would then make it a mandatory parameter

I've pushed a commit such that when the replacementString is set to `""` is either of the above conditions, there is no effect on the original string (default behavior).

File: pkg/ottl/ottlfuncs/func_trim.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	if !ok {
+		return nil, fmt.Errorf("TrimFactory args must be of type *TrimArguments[K]")
+	}
+
+	return trim(args.Target, args.Replacement), nil
+}
+
+func trim[K any](target ottl.StringGetter[K], replacement ottl.Optional[string]) ottl.ExprFunc[K] {
+	return func(ctx context.Context, tCtx K) (any, error) {
+		replacementString := replacement.Get()
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_trim.go:35]:
```suggestion
        replacementString := replacement.Get()
	return func(ctx context.Context, tCtx K) (any, error) {
```

File: pkg/ottl/ottlfuncs/func_trim.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+	if !ok {
+		return nil, fmt.Errorf("TrimFactory args must be of type *TrimArguments[K]")
+	}
+
+	return trim(args.Target, args.Replacement), nil
+}
+
+func trim[K any](target ottl.StringGetter[K], replacement ottl.Optional[string]) ottl.ExprFunc[K] {
+        replacementString := replacement.Get()
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_trim.go:34]: @rnishtala-sumo we do need to still check if `replacement` is empty or not.  If `IsEmpty()` returns true then we need to use a default cut string of `" "`. If `IsEmpty()` returns false then we can use whatever the value of `Get` returns.

File: pkg/ottl/ottlfuncs/func_trim.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+		val, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+		if replacement.IsEmpty() {
+			return strings.Trim(val, " "), nil
+		}
+		return strings.Trim(val, replacementString), nil
+	}
+}
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_trim.go:45]:
```suggestion
func trim[K any](target ottl.StringGetter[K], replacement ottl.Optional[string]) ottl.ExprFunc[K] {
	replacementString := " "
	if !replacement.IsEmpty() {
		replacementString = replacement.Get()
	}
	return func(ctx context.Context, tCtx K) (any, error) {
		val, err := target.Get(ctx, tCtx)
		if err != nil {
			return nil, err
		}
		return strings.Trim(val, replacementString), nil
	}
}
```

**Review by TylerHelmuth — APPROVED** [review]

File: pkg/ottl/ottlfuncs/README.md [hunk 9]
```diff
@@ -1792,9 +1792,21 @@ The `Split` Converter separates a string by the delimiter, and returns an array
...
+### Trim
+
+```Trim(target, Optional[replacement])```
+
+The `Trim` Converter removes the leading and trailing character (default: whitespace).
+
+If the `target` is not a string or does not exist, the `Trim` Converter will return an error.
+
+`target` is a string.
+`replacement` is an optional string representing the character to replace with (default: whitespace).
```
Comment (evan-bradley) [inline pkg/ottl/ottlfuncs/README.md:1804]:
```suggestion
The `Trim` Converter removes the leading and trailing character (default: a space character).

If the `target` is not a string or does not exist, the `Trim` Converter will return an error.

`target` is a string.
`replacement` is an optional string representing the character to replace with (default: a space character).
```
"Whitespace" often also refers to other non-visible characters like tabs, newlines, etc. We should avoid giving that impression here.

**Review by evan-bradley — APPROVED** [review]

**Pull request merged**